tools/istio-iptables/pkg/builder/iptables_builder_impl.go

```diff
@@ -131,12 +132,70 @@ func (rb *IptablesBuilderImpl) BuildV6() [][]string {
 	return rules
 }
 
+func (rb *IptablesBuilderImpl) constructIptablesRestoreContents(tableRulesMap map[string][]string) string {
+	output := ""
+	for table, rules := range tableRulesMap {
+		if len(rules) > 0 {
+			output += fmt.Sprintln("*" + table)
+			for _, r := range rules {
+				output += fmt.Sprintln(r)
+			}
+			output += fmt.Sprintln("COMMIT")
+		}
+	}
+	return output
+}
+
 func (rb *IptablesBuilderImpl) BuildV4Restore() string {
-	//TODO (abhide): implement this
-	return ""
+	tableRulesMap := map[string][]string{
+		constants.FILTER: {},
+		constants.NAT:    {},
+		constants.MANGLE: {},
+	}
+
+	chainTableLookupMap := make(map[string]struct{})
+	for _, r := range rb.rules.rulesv4 {
+		chainTable := fmt.Sprintf("%s:%s", r.chain, r.table)
+		// Create new chain if key: `chainTable` isn't present in map
+		if _, present := chainTableLookupMap[chainTable]; !present {
+			// Ignore chain creation for built-in chains for iptables
+			if _, present := constants.BuiltInChainsMap[r.chain]; !present {
+				tableRulesMap[r.table] = append(tableRulesMap[r.table], fmt.Sprintf("-N %s", r.chain))
+				chainTableLookupMap[chainTable] = struct{}{}
+			}
+		}
+	}
+
+	for _, r := range rb.rules.rulesv4 {
+		tableRulesMap[r.table] = append(tableRulesMap[r.table], strings.Join(r.params, " "))
+	}
+
+	return rb.constructIptablesRestoreContents(tableRulesMap)
 }
 
 func (rb *IptablesBuilderImpl) BuildV6Restore() string {
-	//TODO (abhide): implement this
-	return ""
+	tableRulesMap := map[string][]string{
+		constants.FILTER: {},
+		constants.NAT:    {},
+		constants.MANGLE: {},
+	}
+
+	chainTableLookupMap := make(map[string]struct{})
+	for _, r := range rb.rules.rulesv6 {
+		chainTable := fmt.Sprintf("%s:%s", r.chain, r.table)
```

Comment:
Please factorize this code with `BuildV4Restore() `. The code is identical. This could go into a function taking a `[]*Rule` parameter.